zerver/lib/actions.py

```diff
@@ -2537,32 +2537,42 @@ def decode_email_address(email):
 # performance impact for loading / for users with large numbers of
 # subscriptions, so it's worth optimizing.
 def gather_subscriptions_helper(user_profile):
-    # type: (UserProfile) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]], Dict[int, text_type]]
+    # type: (UserProfile) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]], List[Dict[str, Any]], Dict[int, text_type]]
     sub_dicts = Subscription.objects.select_related("recipient").filter(
         user_profile    = user_profile,
         recipient__type = Recipient.STREAM).values(
         "recipient__type_id", "in_home_view", "color", "desktop_notifications",
         "audible_notifications", "active")
 
     stream_ids = [sub["recipient__type_id"] for sub in sub_dicts]
+    all_streams = get_active_streams(user_profile.realm).select_related(
+        "realm").values("id", "name", "invite_only", "realm_id", \
+        "realm__domain", "email_token", "description")
 
-    stream_dicts = get_active_streams(user_profile.realm).select_related(
-        "realm").filter(id__in=stream_ids).values(
-        "id", "name", "invite_only", "realm_id", "realm__domain", "email_token", "description")
+    stream_dicts = [stream for stream in all_streams if stream['id'] in stream_ids]
```

Comment:
You should switch `stream_ids` to a `set` so that this doesn't take quadratic time